examples/chip-tool/commands/icd/CheckInDelegate.h

```diff
@@ -0,0 +1,35 @@
+/*
+ *   Copyright (c) 2021-2022 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#pragma once
+
+#include <app/icd/client/CheckInHandler.h>
+#include <app/icd/client/DefaultCheckInDelegate.h>
+
+class CheckInDelegate : public chip::app::DefaultCheckInDelegate
+{
+public:
+    void OnCheckInComplete(const chip::app::ICDClientInfo & clientInfo) override;
+
+    void AddExtraDelegate(chip::app::CheckInDelegate * delegate) { mExtraDelegate = delegate; }
```

Comment:
is this AddExtraDelegate necessary? Can we do

void CheckInDelegate::OnCheckInComplete(const chip::app::ICDClientInfo & clientInfo)
{
    chip::app::DefaultCheckInDelegate(clientInfo);
     InteractiveCommand::CheckInComplete() 
}

Reply:
> is this AddExtraDelegate necessary? Can we do
> 
> void CheckInDelegate::OnCheckInComplete(const chip::app::ICDClientInfo & clientInfo) { chip::app::DefaultCheckInDelegate(clientInfo); InteractiveCommand::CheckInComplete(clientInfo) }

No

- InteractiveCommand is not a singleton, we need to pass the instance when calling the callback.

I'm even thinking about adding this function to DefaultCheckInDelegate to support a bit more flexibility.

Reply:
If we go this approach, we should a pool of delegates and not a single a one imo. 
I would add this to the DefaultCheckDelegate and remove this new delegate.

Reply:
Will create another PR for this.

My idea is to make `DefaultCheckInDelegate` unchanged, and create another: `LinkedListCheckInDelegate : public CheckInDelegate`, which provides `AddCheckInDelegate(unique_ptr<CheckInDelegate>&&)`, and we can add DefaultCheckInDelegate to it.

Then the users may create a `LinkedListCheckInDelegate`, and then add `DefaultCheckInDelegate` and some custom check-in delegate to it.

If users don't have their own delegate, then they can just register the `DefaultCheckInDelegate`.

Adding a `AddCheckInDelegate` to `DefaultCheckInDelegate` does not makes sense to me, since it does not looks "default".

Reply:
Let's just relocate the DefaultCheckInDelegate to c++ chip-tool. Obviously different app needs their dedicated CheckInDelegate, android/darwin/python tool...so let's make existing DefaultCheckInDelegate as the dedicated one for DefaultCheckInDelegate.